go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,367 @@
+	if reason != "" {
+		return nil, nil, errors.New(reason)
+	}
+	if table.Keyspace.Sharded {
+		return &Route{
+			PlanID:   SelectScatter,
+			Keyspace: table.Keyspace,
+		}, table, nil
+	}
+	return &Route{
+		PlanID:   SelectUnsharded,
+		Keyspace: table.Keyspace,
+	}, table, nil
+}
+
+// processJoin produces a PlanBuilder subtree and SymbolTable
+// for the given Join. If the left and right nodes can be part
+// of the same route, then it's a RouteBuilder. Otherwise,
+// it's a JoinBuilder.
+func processJoin(join *sqlparser.JoinTableExpr, schema *Schema) (PlanBuilder, *SymbolTable, error) {
+	switch join.Join {
+	case sqlparser.JoinStr, sqlparser.StraightJoinStr, sqlparser.LeftJoinStr:
+	default:
+		// TODO(sougou): better error message.
+		return nil, nil, errors.New("unsupported join")
+	}
+	lplanBuilder, lsymbols, err := processTableExpr(join.LeftExpr, schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	rplanBuilder, rsymbols, err := processTableExpr(join.RightExpr, schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	switch lplanBuilder := lplanBuilder.(type) {
+	case *JoinBuilder:
+		return makeJoinBuilder(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join)
+	case *RouteBuilder:
+		switch rplanBuilder := rplanBuilder.(type) {
+		case *JoinBuilder:
+			return makeJoinBuilder(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join)
+		case *RouteBuilder:
+			return joinRoutes(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join)
+		}
+	}
+	panic("unreachable")
+}
+
+// makeJoinBuilder creates a new JoinBuilder node out of the two builders.
+// This function is called when the two builders cannot be part of
+// the same route.
+func makeJoinBuilder(lplanBuilder PlanBuilder, lsymbols *SymbolTable, rplanBuilder PlanBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr) (PlanBuilder, *SymbolTable, error) {
+	err := lsymbols.Add(rsymbols)
+	if err != nil {
+		return nil, nil, err
+	}
+	isLeft := false
+	if join.Join == sqlparser.LeftJoinStr {
+		isLeft = true
+	}
+	assignOrder(rplanBuilder, lplanBuilder.Order())
+	return &JoinBuilder{
+		IsLeft: isLeft,
+		order:  rplanBuilder.Order(),
+		Left:   lplanBuilder,
+		Right:  rplanBuilder,
+	}, lsymbols, nil
+}
+
+// assignOrder sets the order for the nodes of the tree based on the
+// starting order.
+func assignOrder(planBuilder PlanBuilder, order int) {
+	switch planBuilder := planBuilder.(type) {
+	case *JoinBuilder:
+		assignOrder(planBuilder.Left, order)
+		assignOrder(planBuilder.Right, planBuilder.Left.Order())
+		planBuilder.order = planBuilder.Right.Order()
+	case *RouteBuilder:
+		planBuilder.order = order + 1
+	}
+}
+
+// joinRoutes attempts to join two RouteBuilder objects into one.
+// If it's possible, it produces a joined RouteBuilder.
+// Otherwise, it's a JoinBuilder.
+func joinRoutes(lRouteBuilder *RouteBuilder, lsymbols *SymbolTable, rRouteBuilder *RouteBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr) (PlanBuilder, *SymbolTable, error) {
```

Comment:
Are you planning to also join routes here that are routed to the same shard? Or are you not going to do vindex lookups at this stage?